design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,323 @@
+will register the heartbeats a server receives from other servers. 
+
+Cluster membership is determined using the information provided from the 
+[Kafka Streams](https://kafka.apache.org/20/javadoc/org/apache/kafka/streams/KafkaStreams.html)
+instance and specifically [`StreamsMetadata`](https://kafka.apache.org/20/javadoc/org/apache/kafka/streams/state/StreamsMetadata.html) 
+from which we can obtain the information of the host and port of the instance. A server periodically
+polls for all currently running KS instances (local and remote) and updates the list of remote servers
+it has seen. If no KS streams are currently running, cluster membership is not performed. Moreover,
+this policy depends on the current design where a new KSQL server replays every command in the 
+command topic and hence runs every query.
+
+We want to guarantee 99% uptime. This means that in a 5 minute window, it must take no longer than 
+2 seconds to determine the active server is down and to forward the request to one of the standbys. 
+The heartbeats must be light-weight so that we can send multiple heartbeats per second which will 
+provide us with more datapoints required to implement a well-informed policy for determining when a 
+server is up and when down.
+
+Configuration parameters:
+1) Heartbeat INTERVAL (e.g send heartbeat every 200 ms)
+2) Heartbeat WINDOW (e.g every 2 seconds, count the missed/received heartbeats and determine if server is down/up)
+3) MISSED_THRESHOLD: How many heartbeats in a row constitute a node as down (e.g. 3 missed heartbeats = server is down)
+4) RECEIVED_THRESHOLD: How many heartbeats in a row constitute a node as up (e.g. 2 received heartbeats = server is up)
+
+We will provide sane defaults out-of-box, to achieve a 99% uptime.
+
+Pseudocode for REST endpoint for `/heartbeat`:
+```java
+Map<KsqlServer, List<Long>> receivedHeartbeats;
+
+@POST
+public Response receiveHeartbeat(Request heartbeat) 
+  return Response.ok(processRequest()).build();
+}
+
+private HeartbeatResponse processRequest() {
+  KsqlServer server = request.getServer();
+  receivedHeartbeats.get(server).add(request.getTimestamp());
+}
+```
+
+Additionaly, we will implement a REST endpoint `/clusterStatus` that provides the current status of the cluster
+i.e. which servers are up and which are down (from the viewpoint of the server that receives the
+request).
+
+Pseudocode for REST endpoint for `/clusterStatus`:
+```java
+Map<KsqlServer, Boolean> hostStatus;
+
+@GET
+public Response checkClusterStatus(Request heartbeat) 
+  return Response.ok(processRequest()).build();
+}
+
+private ClusterStatusResponse processRequest() {
+  return hostStatus;
+}
+```
+### Lag reporting
+
+Every server neeeds to periodically broadcast their local lag information. We will implement a new 
+REST endpoint `/reportlag`. The lag information at a server is obtained via 
+`Map<String, Map<Integer, Long>> localLagMap = kafkaStreams.allLocalOffsetLags();`.
```

Comment:
this would have to be run across all streams instances right? Also, the key is a topic name?

Reply:
Yes, this should be run across all streams instances. It's a map of store names to another map of partition to offset lags.